crates/engine/tree/src/download.rs

```diff
@@ -204,6 +228,10 @@ where
             return Poll::Pending;
         }
 
+        if let Some(pending_event) = self.pop_pending_event() {
+            return Poll::Ready(pending_event);
+        }
```

Comment:
maybe emitting here after the if `self.set_buffered_blocks.is_empty() {`
is problematic

we can move this to the start of the poll function because returning poll will ensure poll is called again

Reply:
ok, I wanted to emit the events after the first loops to advance existing requests, but emitting these new request events should happen first :+1: